src/main/java/org/rumbledb/runtime/functions/sequences/general/TreatAsClosure.java

```diff
@@ -2,38 +2,66 @@
+    public TreatAsClosure(SequenceType sequenceType, ErrorCode shouldThrowTreatException, ExceptionMetadata metadata) {
         this.sequenceType = sequenceType;
+        this.errorCode = shouldThrowTreatException;
         this.metadata = metadata;
     }
 
     @Override
     public Boolean call(Item input) throws Exception {
         if (!input.isTypeOf(this.sequenceType.getItemType())) {
-            throw new TreatException(
-                    " "
-                        + input.getDynamicType().toString()
-                        + " cannot be treated as type "
-                        + this.sequenceType.getItemType().toString()
-                        + this.sequenceType.getArity().getSymbol(),
-                    this.metadata
-            );
+            switch (errorCode) {
+                case DynamicTypeTreatErrorCode:
+                    throw new TreatException(
+                            input.getDynamicType().toString()
+                                + " cannot be treated as type "
+                                + this.sequenceType.getItemType().toString()
+                                + this.sequenceType.getArity().getSymbol(),
+                            this.metadata
+                    );
+                case UnexpectedTypeErrorCode:
+                    throw new UnexpectedTypeException(
+                            input.getDynamicType().toString()
+                                + " is not expected here. The expected type is "
+                                + this.sequenceType.getItemType().toString()
+                                + this.sequenceType.getArity().getSymbol(),
+                            this.metadata
+                    );
+                default:
+                    throw new OurBadException("Unexpected error code in treat as iterator.", this.metadata);
+            }
         }
         if (this.sequenceType.isEmptySequence()) {
-            throw new TreatException(
-                    input.getDynamicType().toString()
-                        + " cannot be treated as type empty-sequence()",
-                    this.metadata
-            );
+            switch (errorCode) {
+                case DynamicTypeTreatErrorCode:
+                    throw new TreatException(
+                            input.getDynamicType().toString()
+                                + " cannot be treated as type empty-sequence()",
+                            this.metadata
+                    );
+                case UnexpectedTypeErrorCode:
+                    throw new UnexpectedTypeException(
+                            input.getDynamicType().toString()
+                                + " is not expected here. The expected type is empty-sequence()",
+                            this.metadata
+                    );
+                default:
+                    throw new OurBadException("Unexpected error code in treat as iterator.", this.metadata);
+            }
```

Comment:
The new messages look very clear now. Shall we use the same messages in TreatIterator.java for consistency?